### PR TITLE
Modify logic for final model training

### PR DESCRIPTION
**Pull request by fraimondo**
So far, when the user requested the final model, after calling scikit-learn's `cross_validate`, julearn was fiting the model again, on the full training data.

The main issue is when using joblib to parallelize, there was a call for each outer CV fold and once it was done, the main process will fit the final model. With enough resources, this is suboptimal, as one might want to fit the final model at the same time of the individual folds.

This PR changes the internal logic so the effect is the same, but the fiting happens at a different time. The idea is to add an "extra" fold in the CV object which includes the whole dataset. After the call to `cross_validate` is done, we remove the last entry and use this as the final model, obtaining the same output, but allowing the user to use joblib to parallelise together across CV folds and the final model.

### CODE REVIEW

File: julearn/model_selection/utils.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+# Authors: Federico Raimondo <f.raimondo@fz-juelich.de>
+# License: AGPL
+
+from sklearn.model_selection import check_cv as sk_check_cv
+
+from .final_model_cv import _JulearnFinalModelCV
+
+
+def check_cv(cv=5, classifier=False, include_final_model=False):
```
Comment (synchon) [inline julearn/model_selection/utils.py:11]: Missing type annotations

File: julearn/model_selection/utils.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    Parameters
+    ----------
+    cv : int, str or cross-validation generator | None
+        Cross-validation splitting strategy to use for model evaluation.
+
+        Options are:
+
+        * None: defaults to 5-fold
+        * int: the number of folds in a `(Stratified)KFold`
+        * CV Splitter (see scikit-learn documentation on CV)
```
Comment (synchon) [inline julearn/model_selection/utils.py:23]: Maybe a link here?

File: julearn/model_selection/tests/test_final_model_cv.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+"""Provides tests for the final model CV."""
+
+# Authors: Federico Raimondo <f.raimondo@fz-juelich.de>
+# License: AGPL
```
Comment (synchon) [inline julearn/model_selection/tests/test_final_model_cv.py:4]: Nitpick: empty line after this

File: julearn/model_selection/final_model_cv.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
+"""CV Wrapper that includes a fold with all the data."""
+
+# Authors: Federico Raimondo <f.raimondo@fz-juelich.de>
+# License: AGPL
```
Comment (synchon) [inline julearn/model_selection/final_model_cv.py:4]: Nitpick: empty line after this

**Review by synchon — CHANGES_REQUESTED** [review]

File: docs/changes/newsfragments/293.enh [hunk 5]
```diff
@@ -0,0 +1 @@
+Change the internal logic of `:func:.run_cross_validation` to optimise joblib calls by `Fede Raimondo`_
```
Comment (synchon) [inline docs/changes/newsfragments/293.enh:1]: I believe

```
`:func:.run_cross_validation`
```

needs to be

```
:func:`.run_cross_validation`
```
?

File: julearn/model_selection/final_model_cv.py (hunk 4 above)
Comment (fraimondo) [reply julearn/model_selection/final_model_cv.py:4]: done

File: docs/changes/newsfragments/293.enh (hunk 5 above)
Comment (fraimondo) [reply docs/changes/newsfragments/293.enh:1]: yes

File: julearn/model_selection/utils.py (hunk 2 above)
Comment (fraimondo) [reply julearn/model_selection/utils.py:23]: That's a copy-paste from other docs. It's not easy to link to that scikit-learn section

File: julearn/model_selection/tests/test_final_model_cv.py (hunk 3 above)
Comment (fraimondo) [reply julearn/model_selection/tests/test_final_model_cv.py:4]: done

File: julearn/model_selection/utils.py (hunk 1 above)
Comment (fraimondo) [reply julearn/model_selection/utils.py:11]: done

**Pull request merged**